fatjdbc/src/test/java/com/qubole/quark/fatjdbc/test/MaterializedViewJoinTest.java

```diff
@@ -0,0 +1,129 @@
+    connInfo.setProperty("db:" +
+        MaterializedViewJoinTest.class.
+            getResource("/HiveUDFModel.json").getPath(), "");
+    connInfo.setProperty("schemaFactory",
+        "com.qubole.quark.catalog.db.SchemaFactory");
+    connInfo.setProperty("encryptionKey", "key");
+
+    dbConnection = DriverManager.getConnection(dbSchemaUrl, connInfo);
+
+    Statement stmt = dbConnection.createStatement();
+    String sql = "insert into data_sources(name, type, url, ds_set_id, datasource_type) values "
+        + "('H2', 'H2', '" + url + "', 1, 'JDBC'); insert into jdbc_sources (id, "
+        + "username, password) values(1, '" + username + "', '" + password + "');"
+        + "update ds_sets set default_datasource_id = 1 where id = 1;"
+        + "insert into partitions(`name`, `description`, `cost`, `query`, `ds_set_id`,"
+        + " `destination_id`,`schema_name`, `table_name`)"
+        + " VALUES('test_hist_part', 'Test History Partition', 0, "
+        + "'select * from h2.public.test_hist as h where h.timeout > 300',"
+        + " 1, 1, 'PUBLIC', 'TEST_HIST_PARTITION');";
+
+    stmt.execute(sql);
+    stmt.close();
+  }
+
+  public static void setUpClass(String dbUrl) throws Exception {
+    Class.forName("com.qubole.quark.fatjdbc.QuarkDriver");
+    Class.forName("org.h2.Driver");
+
+    Properties connInfo = new Properties();
+    connInfo.setProperty("url", dbUrl);
+    connInfo.setProperty("user", "sa");
+    connInfo.setProperty("password", "");
+
+    h2Connection = DriverManager.getConnection(dbUrl, connInfo);
+
+    Statement stmt = h2Connection.createStatement();
+    java.net.URL url = MaterializedViewJoinTest.class.getResource("/HiveUDF.sql");
+    java.nio.file.Path resPath = java.nio.file.Paths.get(url.toURI());
+    String sql = new String(java.nio.file.Files.readAllBytes(resPath), "UTF8");
+
+    stmt.execute(sql);
+    stmt.close();
+  }
+
+  @AfterClass
+  public static void tearDownClass() throws SQLException {
+    h2Connection.close();
+  }
+
+  @Test
+  public void testComplexQuery() throws Exception {
+    String sql = "select\n" +
+        "qh.created_at as dt, \n" +
+        "count(qh.id) as num_queries\n" +
+        "\n" +
+        "from test_hist qh\n" +
+        "join uinfo ui\n" +
+        "  on qh.qbol_user_id = ui.qu_id\n" +
+        "join acc externals\n" +
+        "  on externals.id = ui.a_id\n" +
+        "\n" +
+        "where qh.timeout >= 350\n" +
+        "and command_type = 'HiveCommand'\n" +
+        "and qlog like '%\\\"HIVE_VERSION\\\":\\\"1.2\\\"%'\n" +
+        "and customer_name like 'amogh'\n" +
+        "\n" +
+        "group by \n" +
+        "qh.created_at\n" +
+        "\n" +
+        "order by dt asc";
+    SqlQueryParser parser = new SqlQueryParser(connInfo);
+    final SqlQueryParser.SqlQueryParserResult result = parser.parse(sql);
+    assertEquals(result.getParsedSql(), "SELECT \"t0\".\"CREATED_AT\", COUNT(\"t0\".\"ID\") AS \"NUM_QUERIES\" FROM "
```

Comment:
Good to confirm it by running an equivalent query in Hive.

On Thu, Mar 31, 2016 at 11:52 AM amoghmargoor <notifications@github.com>
wrote:

> In
> fatjdbc/src/test/java/com/qubole/quark/fatjdbc/test/MaterializedViewJoinTest.java
> <https://github.com/qubole/quark/pull/61#discussion_r58005458>:
>
> > +        "  on qh.qbol_user_id = ui.qu_id\n" +
> > +        "join acc externals\n" +
> > +        "  on externals.id = ui.a_id\n" +
> > +        "\n" +
> > +        "where qh.timeout >= 350\n" +
> > +        "and command_type = 'HiveCommand'\n" +
> > +        "and qlog like '%\\\"HIVE_VERSION\\\":\\\"1.2\\\"%'\n" +
> > +        "and customer_name like 'amogh'\n" +
> > +        "\n" +
> > +        "group by \n" +
> > +        "qh.created_at\n" +
> > +        "\n" +
> > +        "order by dt asc";
> > +    SqlQueryParser parser = new SqlQueryParser(connInfo);
> > +    final SqlQueryParser.SqlQueryParserResult result = parser.parse(sql);
> > +    assertEquals(result.getParsedSql(), "SELECT \"t0\".\"CREATED_AT\", COUNT(\"t0\".\"ID\") AS \"NUM_QUERIES\" FROM "
>
> Ohh hive has some issues with join on subqueries ? Will verify if it can
> run join on subqueries.
>
> The reason sub query was introduced was because our new rule kicks in only
> after filter clause has been pushed down the join on to table scan.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/qubole/quark/pull/61/files/70ed3407badc788714b527b30b24816a9509b6fe#r58005458>
>